### PR TITLE
8271884: G1CH::_expand_heap_after_alloc_failure is no longer needed

### PR DESCRIPTION
**Pull request by walulyai**
Hi,

Please review this change to remove _expand_heap_after_alloc_failure. Checks for old_is_full(), survivor_is_full(), and G1CollectedHeap::has_more_regions() should eliminate repeated expansion attempts.

Testing; tier 1

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271884](https://bugs.openjdk.java.net/browse/JDK-8271884): G1CH::_expand_heap_after_alloc_failure is no longer needed

### Reviewers
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to df0ef8804fa4b32c1d9edd8280bc15be0be44f74
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5030/head:pull/5030` \
`$ git checkout pull/5030`

Update a local copy of the PR: \
`$ git checkout pull/5030` \
`$ git pull https://git.openjdk.java.net/jdk pull/5030/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5030`

View PR using the GUI difftool: \
`$ git pr show -t 5030`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5030.diff">https://git.openjdk.java.net/jdk/pull/5030.diff</a>

</details>

### CODE REVIEW

**Review by kimbarrett — APPROVED** [review]
Change looks good.

Testing should include verifying there isn't large amounts of output when
the appropriate logging is enabled and there are expansion failures. I filed
the RFE based on code-examination; it would be a pain if I missed something
and this flag actually is still needed.

File: src/hotspot/share/gc/g1/g1CollectedHeap.cpp [hunk 1]
```diff
@@ -164,11 +164,10 @@ HeapRegion* G1CollectedHeap::new_region(size_t word_size,
 
   HeapRegion* res = _hrm.allocate_free_region(type, node_index);
 
-  if (res == NULL && do_expand && _expand_heap_after_alloc_failure) {
+  if (res == NULL && do_expand) {
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1CollectedHeap.cpp:167]: Ivan and me were looking into any adverse impact when removing this, in particular the expansion being called very often without `_expand_heap_after_alloc_failure`.
There is indeed an issue with convoying in both the survivor and old gen region allocation; the code that ultimately calls down here looks like the following:
```
  if (result == NULL && !old_is_full()) {
    MutexLocker x(FreeList_lock, Mutex::_no_safepoint_check_flag);
    result = old_gc_alloc_region()->attempt_allocation_locked(min_word_size,
                                                              desired_word_size,
                                                              actual_word_size);
    if (result == NULL) {
      set_old_full();
    }
  }
  ```
(likewise for survivor allocation)
After the first check to `old_is_full()` we get the `FreeList_lock` and after that call down to the allocation. This is actually the problem: multiple threads can be waiting for the lock trying to allocate, and all of these waiters will be able to try to expand without that `_expand_heap_after_alloc_failure` flag.
The solution is to check `old_is_full()` after grabbing the lock again - which is actually better than only finding out here quite late that there is nothing to allocate.

**Review by tschatzl — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/g1/g1CollectedHeap.cpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1CollectedHeap.cpp:167]: We should probably check if there is a similar convoying effect for mutator allocation, possibly in a separate CR.

Comment (walulyai) [reply src/hotspot/share/gc/g1/g1CollectedHeap.cpp:167]: Thanks @tschatzl  for the detailed explanation. Fixed!

File: src/hotspot/share/gc/g1/g1Allocator.cpp [hunk 2]
```diff
@@ -272,11 +274,13 @@ HeapWord* G1Allocator::old_attempt_allocation(size_t min_word_size,
                                                                actual_word_size);
   if (result == NULL && !old_is_full()) {
     MutexLocker x(FreeList_lock, Mutex::_no_safepoint_check_flag);
-    result = old_gc_alloc_region()->attempt_allocation_locked(min_word_size,
-                                                              desired_word_size,
-                                                              actual_word_size);
-    if (result == NULL) {
-      set_old_full();
+    if (!old_is_full()) {
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1Allocator.cpp:277]: Please add a short comment about the reasons for the re-check here (and above)

**Review by tschatzl — APPROVED** [review]
Lgtm, please add the suggested comments.

**Review by tschatzl — APPROVED** [review]

**Pull request closed without merging**